pkg/status/singletonstatus.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+type singletonState struct {
+	sync.RWMutex
+	wObjSync map[util.ObjectIdentifier]bool
+}
+
+func (ss *singletonState) get(id util.ObjectIdentifier) bool {
+	ss.Lock()
+	defer ss.Unlock()
+
+	return ss.wObjSync[id]
+}
+
+func (ss *singletonState) set(id util.ObjectIdentifier) {
+	ss.Lock()
+	defer ss.Unlock()
+
+	ss.wObjSync[id] = true
+}
+
+func (ss *singletonState) unset(id util.ObjectIdentifier) {
+	ss.Lock()
+	defer ss.Unlock()
+
+	ss.wObjSync[id] = false
+}
+
+func (ss *singletonState) addIfNotExist(id util.ObjectIdentifier) {
+	ss.Lock()
+	defer ss.Unlock()
+
+	if _, ok := ss.wObjSync[id]; !ok {
+		ss.wObjSync[id] = false
+	}
+}
+
+func (ss *singletonState) delete(id util.ObjectIdentifier) {
+	ss.Lock()
+	defer ss.Unlock()
+
+	delete(ss.wObjSync, id)
+}
+
+func (c *Controller) buildSingletonStateAndOptionallyReconcile(ctx context.Context, reconcile bool) error {
+	logger := klog.FromContext(ctx)
+	logger.V(2).Info("Building the desired state for singleton statuses")
+
+	for wObjIdentifier := range c.singletonState.wObjSync {
+		gvr := schema.GroupVersionResource{Group: wObjIdentifier.GVK.Group, Version: wObjIdentifier.GVK.Version, Resource: wObjIdentifier.Resource}
+		lister, found := c.listers.Get(gvr)
+		if !found {
+			return fmt.Errorf("could not get lister for gvr: %s", gvr)
+		}
+		_, err := lister.ByNamespace(wObjIdentifier.ObjectName.Namespace).Get(wObjIdentifier.ObjectName.Name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				c.singletonState.delete(wObjIdentifier)
+			}
+			return err
+		}
+	}
+
+	allBdgs, err := c.wdsKsClient.ControlV1alpha1().Bindings().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, bdg := range allBdgs.Items {
+		logger.V(2).Info("Got Binding while building singleton state", "name", bdg.Name)
```

Comment:
I think this should be at a numerically higher `V` level, it is a pretty low-level debugging detail.

Reply:
Yes, explained [here](https://github.com/kubestellar/kubestellar/pull/2285#discussion_r1709546859).